pil-analyzer/src/condenser.rs

```diff
@@ -674,45 +678,190 @@ fn to_expr<T: Clone>(value: &Value<'_, T>) -> AlgebraicExpression<T> {
+            )));
+        }
+        *kind = expected_kind;
+    }
+
+    Ok(FunctionValueDefinition::Expression(TypedExpression {
+        e,
+        type_scheme: None,
+    }))
+}
+
+fn try_closure_to_expression<T: FieldElement>(
+    closure: &evaluator::Closure<'_, T>,
+) -> Result<Expression, EvalError> {
+    if !closure.type_args.is_empty() {
         return Err(EvalError::TypeError(
             "Lambda expression must not have type arguments.".to_string(),
         ));
     }
-    if !lambda.outer_var_references.is_empty() {
-        return Err(EvalError::TypeError(format!(
-            "Lambda expression must not reference outer variables: {lambda}"
-        )));
-    }
-    if lambda.kind != FunctionKind::Pure && lambda.kind != expected_kind {
-        return Err(EvalError::TypeError(format!(
-            "Expected {expected_kind} lambda expression but got {}.",
-            lambda.kind
-        )));
-    }
 
-    let mut lambda = (*lambda).clone();
-    lambda.kind = expected_kind;
+    let var_height = closure.environment.len() as u64;
+    let outer_var_refs =
+        outer_var_refs(var_height, &closure.lambda.body).collect::<BTreeMap<_, _>>();
 
-    Ok(FunctionValueDefinition::Expression(TypedExpression {
-        e: Expression::LambdaExpression(source.clone(), lambda),
-        type_scheme: None,
-    }))
+    let mut lambda = (*closure.lambda).clone();
+
+    compact_var_refs(
+        &mut lambda.body,
+        &outer_var_refs.keys().copied().collect::<Vec<_>>(),
+        var_height,
+    );
+
+    let statements = outer_var_refs
+        .into_iter()
+        .map(|(v_id, name)| {
+            let value = Some(try_value_to_expression(
+                closure.environment[v_id as usize].as_ref(),
+            )?);
+
+            Ok(LetStatementInsideBlock {
+                pattern: Pattern::Variable(SourceRef::unknown(), name.clone()),
+                // We do not know the type.
+                ty: None,
+                value,
+            }
+            .into())
+        })
+        .collect::<Result<Vec<_>, _>>()?;
+    let e = Expression::LambdaExpression(SourceRef::unknown(), lambda);
+
+    Ok(if statements.is_empty() {
+        e
+    } else {
+        BlockExpression {
+            statements,
+            expr: Some(Box::new(e)),
+        }
+        .into()
+    })
+}
+
+fn outer_var_refs(environment_size: u64, e: &Expression) -> impl Iterator<Item = (u64, &String)> {
+    e.all_children().filter_map(move |e| {
+        if let Expression::Reference(_, Reference::LocalVar(id, name)) = e {
+            (*id < environment_size).then_some((*id, name))
+        } else {
+            None
+        }
+    })
+}
+
+/// Updates local variable IDs to be compact.
+fn compact_var_refs(e: &mut Expression, referenced_outer_vars: &[u64], environment_size: u64) {
+    e.children_mut().for_each(|e| {
+        if let Expression::Reference(_, Reference::LocalVar(id, _name)) = e {
```

Comment:
```suggestion
        if let Expression::Reference(_, Reference::LocalVar(id, _)) = e {
```

Reply:
Also below it turns out that some of these references are not to local variables but to captured variable, so maybe `LocalVar` is not the right naming here?

Reply:
The distinction is between a local variable and a globally accessible path (named PolynomialReference for historical reasons, should just be called "SymbolReference" or something).
I think it makes it more complicated to distinguish between variables that were captured and those that were not.